File: 106.construct-binary-tree-from-inorder-and-postorder-traversal.java
/*
 * @lc app=leetcode id=106 lang=java
 *
 * [106] Construct Binary Tree from Inorder and Postorder Traversal
 */

// @lc code=start

import java.util.Arrays;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */

class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder.length == 0)
            return null;
        if (inorder.length == 1)
            return new TreeNode(inorder[0]);
        TreeNode root = new TreeNode(postorder[postorder.length - 1]);
        int idx = 0;
        for (int ix = 0; ix != inorder.length; ++ix) {
            if (inorder[ix] == postorder[postorder.length - 1]) {
                idx = ix;
            }
        }
        int[] leftInorder = Arrays.copyOfRange(inorder, 0, idx);
        int[] rightInorder = Arrays.copyOfRange(inorder, idx + 1, inorder.length);
        int[] leftPostorder = Arrays.copyOfRange(postorder, 0, leftInorder.length);
        int[] rightPostorder = Arrays.copyOfRange(postorder, leftPostorder.length, postorder.length - 1);
        root.left = buildTree(leftInorder, leftPostorder);
        root.right = buildTree(rightInorder, rightPostorder);
        return root;
    }
}
// @lc code=end

File: 106.construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/*
 * @lc app=leetcode id=106 lang=cpp
 *
 * [106] Construct Binary Tree from Inorder and Postorder Traversal
 */

// @lc code=start

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:

    using iterator = vector<int>::const_iterator;

    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
        if (inorder.empty()) return nullptr;
        return buildTree(inorder.cbegin(), inorder.cend(), postorder.cbegin(), postorder.cend());
    }

    TreeNode *buildTree(
        iterator inorder_begin, iterator inorder_end,
        iterator postorder_begin, iterator postorder_end) {
        auto root = new TreeNode { *(postorder_end - 1) };
        if (distance(postorder_begin, postorder_end) == 1) return root;

        auto pos = find(inorder_begin, inorder_end, *(postorder_end - 1));
        auto left_cnt = distance(inorder_begin, pos);

        auto left_inorder_begin = inorder_begin, left_inorder_end = pos,
            right_inorder_begin = pos + 1, right_inorder_end = inorder_end;
        auto left_postorder_begin = postorder_begin, left_postorder_end = left_postorder_begin + left_cnt,
            right_postorder_begin = left_postorder_end, right_postorder_end = postorder_end - 1;

        if (distance(left_inorder_begin, left_inorder_end) != 0) {
            root->left = buildTree(left_inorder_begin, left_inorder_end, left_postorder_begin, left_postorder_end);
        }
        if (distance(right_inorder_begin, right_inorder_end) != 0) {
            root->right = buildTree(right_inorder_begin, right_inorder_end, right_postorder_begin, right_postorder_end);
        }
        return root;
    }
};
// @lc code=end